lib/ansible/modules/packaging/os/package_facts.py

```diff
@@ -140,94 +145,236 @@
+        if rc != 0 or err:
+            raise Exception("Unable to list packages rc=%s : %s" % (rc, err))
+        return out.splitlines()
+
+    def get_package_details(self, package):
+
+        pkg = dict(zip(self.atoms, package.split('\t')))
+
+        if 'arch' in pkg:
             try:
-                dummy = __import__(manager_lib)
-                manager = manager_lib
-                break
-            except ImportError:
+                pkg['arch'] = pkg['arch'].split(':')[2]
+            except IndexError:
                 pass
 
-        # FIXME: add more detection methods
-    try:
-        if manager == "rpm":
-            packages = rpm_package_list()
-        elif manager == "apt":
-            packages = apt_package_list()
-        else:
-            if manager:
-                results['msg'] = 'Unsupported package manager: %s' % manager
-                results['skipped'] = True
+        if 'automatic' in pkg:
+            pkg['automatic'] = bool(pkg['automatic'])
+
+        if 'category' in pkg:
+            pkg['category'] = pkg['category'].split('/', 1)[0]
+
+        if 'version' in pkg:
+            if ',' in pkg['version']:
+                pkg['version'], pkg['port_epoch'] = pkg['version'].split(',', 1)
             else:
-                module.fail_json(msg='Could not detect supported package manager')
-    except Exception as e:
-        from traceback import format_tb
-        module.fail_json(msg='Failed to retrieve packages: %s' % to_text(e), exception=format_tb(sys.exc_info()[2]))
+                pkg['port_epoch'] = 0
+
+            if '_' in pkg['version']:
+                pkg['version'], pkg['revision'] = pkg['version'].split('_', 1)
+            else:
+                pkg['revision'] = 0
+
+        if 'vital' in pkg:
+            pkg['vital'] = bool(pkg['vital'])
+
+        return pkg
+
+
+class PORTAGE(CLIMgr):
+
+    name = 'portage'
+    CLI = 'qlist'
+    atoms = ['category', 'name', 'version', 'ebuild_revision', 'slots', 'prefixes', 'sufixes']
+
+    def list_installed(self):
+        rc, out, err = self.m.run_command(' '.join([self.cli, '-Iv', '|', 'xargs', '-n', '1024', 'qatom']), use_unsafe_shell=True)
+        if rc != 0 or err:
+            raise RuntimeError("Unable to list packages rc=%s : %s" % (rc, to_native(err)))
+        return out.splitlines()
+
+    def get_package_details(self, package):
+        return dict(zip(self.atoms, package.split()))
+
+
+class RPM(LibMgr):
+
+    name = 'rpm'
+    LIB = 'rpm'
+
+    def list_installed(self):
+        return self.lib.TransactionSet().dbMatch()
+
+    def get_package_details(self, package):
+        return dict(name=package[self.lib.RPMTAG_NAME],
+                    version=package[self.lib.RPMTAG_VERSION],
+                    release=package[self.lib.RPMTAG_RELEASE],
+                    epoch=package[self.lib.RPMTAG_EPOCH],
+                    arch=package[self.lib.RPMTAG_ARCH],)
+
+
+def main():
+
+    me = sys.modules[__name__]
```

Comment:
You don't need this.  Use ```globals().get([..])``` instead.

Reply:
@abadger you can suggest changes by hitting Ctrl+G ;)
```suggestion
    me = globals().get(__name__)
```

Reply:
I'm unsure what you mean here, can you expand?

Reply:
here I meant:
* remove this line ```me = sys.modules[__name__]``` 
* When you retrieve the manager do it like this:             ```manager = globals().get(pkgmgr.upper())```

However, we shouldn't have to do either of these if you use ```get_all_subclasses()``` below.  We can do it in a better way then as PKG_MANAGERS will be a dict mapping directly to the class we need.

Reply:
dont need this anymore due to pkg_mangers updates below